File: go/conformance_fixtures_test.go
package ampsdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

type transitionCase struct {
	Name                 string `json:"name"`
	From                 string `json:"from"`
	To                   string `json:"to"`
	PendingHumanApproval bool   `json:"pending_human_approval"`
	ActorIsParticipant   bool   `json:"actor_is_participant"`
	Valid                bool   `json:"valid"`
}

type transitionFixture struct {
	Cases []transitionCase `json:"cases"`
}

func fixturePath(parts ...string) string {
	base := []string{"..", "spec", "fixtures"}
	base = append(base, parts...)
	return filepath.Join(base...)
}

func readJSONFixture(t *testing.T, parts ...string) []byte {
	t.Helper()
	payload, err := os.ReadFile(fixturePath(parts...))
	if err != nil {
		t.Fatalf("read fixture %v: %v", parts, err)
	}
	return payload
}

func assertJSONRoundTripEqual(t *testing.T, fixture []byte, model any) {
	t.Helper()
	actualBytes, err := json.Marshal(model)
	if err != nil {
		t.Fatalf("marshal model: %v", err)
	}

	var expectedJSON any
	if err := json.Unmarshal(fixture, &expectedJSON); err != nil {
		t.Fatalf("unmarshal expected json: %v", err)
	}

	var actualJSON any
	if err := json.Unmarshal(actualBytes, &actualJSON); err != nil {
		t.Fatalf("unmarshal actual json: %v", err)
	}

	if !reflect.DeepEqual(expectedJSON, actualJSON) {
		t.Fatalf("json mismatch\nexpected: %#v\nactual:   %#v", expectedJSON, actualJSON)
	}
}

func TestConformanceHTTPFixturesRoundTrip(t *testing.T) {
	t.Parallel()

	registerReqBytes := readJSONFixture(t, "http", "register_agent_request.json")
	var registerReq RegisterAgentRequest
	if err := json.Unmarshal(registerReqBytes, &registerReq); err != nil {
		t.Fatalf("unmarshal register request fixture: %v", err)
	}
	assertJSONRoundTripEqual(t, registerReqBytes, registerReq)

	createProfileReqBytes := readJSONFixture(t, "http", "create_profile_request.json")
	var createProfileReq CreateProfileRequest
	if err := json.Unmarshal(createProfileReqBytes, &createProfileReq); err != nil {
		t.Fatalf("unmarshal create profile request fixture: %v", err)
	}
	assertJSONRoundTripEqual(t, createProfileReqBytes, createProfileReq)

	createNegotiationReqBytes := readJSONFixture(t, "http", "create_negotiation_request.json")
	var createNegotiationReq CreateNegotiationRequest
	if err := json.Unmarshal(createNegotiationReqBytes, &createNegotiationReq); err != nil {
		t.Fatalf("unmarshal create negotiation request fixture: %v", err)
	}
	assertJSONRoundTripEqual(t, createNegotiationReqBytes, createNegotiationReq)
}

func TestConformanceHTTPFixturesResponseDeserialization(t *testing.T) {
	t.Parallel()

	registerRespBytes := readJSONFixture(t, "http", "register_agent_response.json")
	var registerResp RegisterAgentResponse
	if err := json.Unmarshal(registerRespBytes, &registerResp); err != nil {
		t.Fatalf("unmarshal register response fixture: %v", err)
	}
	if registerResp.AgentID != "11111111-1111-1111-1111-111111111111" {
		t.Fatalf("unexpected agent id: %s", registerResp.AgentID)
	}

	discoveryRespBytes := readJSONFixture(t, "http", "discovery_response.json")
	var discoveryResp DiscoveryResponse
	if err := json.Unmarshal(discoveryRespBytes, &discoveryResp); err != nil {
		t.Fatalf("unmarshal discovery response fixture: %v", err)
	}
	if len(discoveryResp.Candidates) != 1 || discoveryResp.Candidates[0].Tier != "trusted" {
		t.Fatalf("unexpected discovery fixture parse: %+v", discoveryResp)
	}

	approvalStatusBytes := readJSONFixture(t, "http", "approval_status_response.json")
	var approvalStatus ApprovalStatus
	if err := json.Unmarshal(approvalStatusBytes, &approvalStatus); err != nil {
		t.Fatalf("unmarshal approval status fixture: %v", err)
	}
	if !approvalStatus.Pending || approvalStatus.CurrentState != "DISCLOSED" {
		t.Fatalf("unexpected approval status fixture parse: %+v", approvalStatus)
	}
}

func TestConformanceStateTransitionFixtures(t *testing.T) {
	t.Parallel()

	payload := readJSONFixture(t, "state", "transition_cases.json")
	var fixture transitionFixture
	if err := json.Unmarshal(payload, &fixture); err != nil {
		t.Fatalf("unmarshal transition fixture: %v", err)
	}

	for _, c := range fixture.Cases {
		c := c
		t.Run(c.Name, func(t *testing.T) {
			t.Parallel()
			if got := validateTransition(c); got != c.Valid {
				t.Fatalf("expected %v, got %v", c.Valid, got)
			}
		})
	}
}

func validateTransition(c transitionCase) bool {
	terminal := map[string]bool{
		"WITHDRAWN":   true,
		"EXPIRED":     true,
		"REJECTED":    true,
		"BLOCKED":     true,
		"SAFETY_HOLD": true,
	}
	progression := []string{"DISCOVERY", "INTEREST", "MUTUAL", "DISCLOSING", "DISCLOSED", "MEETING", "ACTIVE"}
	humanGated := map[string]bool{
		"MUTUAL->DISCLOSING":  true,
		"DISCLOSED->MEETING": true,
	}

	if terminal[c.From] {
		return false
	}

	if c.To == "WITHDRAWN" {
		return c.ActorIsParticipant
	}

	if terminal[c.To] {
		return false
	}

	expected := ""
	for idx, state := range progression {
		if state == c.From {
			if idx+1 < len(progression) {
				expected = progression[idx+1]
			}
			break
		}
	}
	if c.To != expected {
		return false
	}

	if humanGated[c.From+"->"+c.To] && !c.PendingHumanApproval {
		return false
	}

	return true
}

File: java/src/test/java/com/bonsai/amp/sdk/ConformanceFixturesTest.java
package com.bonsai.amp.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.bonsai.amp.sdk.model.ApprovalStatus;
import com.bonsai.amp.sdk.model.CreateNegotiationRequest;
import com.bonsai.amp.sdk.model.CreateProfileRequest;
import com.bonsai.amp.sdk.model.DiscoveryResponse;
import com.bonsai.amp.sdk.model.RegisterAgentRequest;
import com.bonsai.amp.sdk.model.RegisterAgentResponse;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.Test;

class ConformanceFixturesTest {
  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().registerModule(new JavaTimeModule());
  private static final List<String> PROGRESSION =
      List.of("DISCOVERY", "INTEREST", "MUTUAL", "DISCLOSING", "DISCLOSED", "MEETING", "ACTIVE");
  private static final Set<String> TERMINAL =
      Set.of("WITHDRAWN", "EXPIRED", "REJECTED", "BLOCKED", "SAFETY_HOLD");

  @Test
  void requestFixturesRoundTrip() throws Exception {
    assertRoundTrip("http/register_agent_request.json", RegisterAgentRequest.class);
    assertRoundTrip("http/create_profile_request.json", CreateProfileRequest.class);
    assertRoundTrip("http/create_negotiation_request.json", CreateNegotiationRequest.class);
  }

  @Test
  void responseFixturesDeserialize() throws Exception {
    RegisterAgentResponse registerAgent =
        readFixtureValue("http/register_agent_response.json", RegisterAgentResponse.class);
    assertEquals("11111111-1111-1111-1111-111111111111", registerAgent.agentId());
    assertEquals("pending_human_verify", registerAgent.status());

    DiscoveryResponse discovery = readFixtureValue("http/discovery_response.json", DiscoveryResponse.class);
    assertEquals(1, discovery.candidates().size());
    assertEquals("trusted", discovery.candidates().get(0).tier());

    ApprovalStatus approval = readFixtureValue("http/approval_status_response.json", ApprovalStatus.class);
    assertTrue(approval.pending());
    assertEquals("DISCLOSED", approval.currentState());
  }

  @Test
  void transitionFixturesMatchCanonicalRules() throws Exception {
    TransitionFixture fixture = readFixtureValue("state/transition_cases.json", TransitionFixture.class);
    for (TransitionCase transitionCase : fixture.cases()) {
      boolean actual =
          validateTransition(
              transitionCase.from(),
              transitionCase.to(),
              transitionCase.pendingHumanApproval(),
              transitionCase.actorIsParticipant());
      assertEquals(actual, transitionCase.valid(), transitionCase.name());
    }
  }

  private static boolean validateTransition(
      String current, String target, boolean pendingHumanApproval, boolean actorIsParticipant) {
    if (TERMINAL.contains(current)) {
      return false;
    }

    if ("WITHDRAWN".equals(target)) {
      return actorIsParticipant;
    }

    if (Set.of("EXPIRED", "REJECTED", "BLOCKED", "SAFETY_HOLD").contains(target)) {
      return false;
    }

    String expected = nextState(current);
    if (!target.equals(expected)) {
      return false;
    }

    if ((("MUTUAL".equals(current) && "DISCLOSING".equals(target))
            || ("DISCLOSED".equals(current) && "MEETING".equals(target)))
        && !pendingHumanApproval) {
      return false;
    }

    return true;
  }

  private static String nextState(String current) {
    int idx = PROGRESSION.indexOf(current);
    if (idx < 0 || idx == PROGRESSION.size() - 1) {
      return null;
    }
    return PROGRESSION.get(idx + 1);
  }

  private static <T> void assertRoundTrip(String fixture, Class<T> modelType) throws Exception {
    JsonNode expected = readFixtureTree(fixture);
    T model = OBJECT_MAPPER.treeToValue(expected, modelType);
    JsonNode actual = OBJECT_MAPPER.valueToTree(model);
    assertEquals(expected, actual);
  }

  private static JsonNode readFixtureTree(String fixture) throws IOException {
    return OBJECT_MAPPER.readTree(fixturePath(fixture).toFile());
  }

  private static <T> T readFixtureValue(String fixture, Class<T> modelType) throws IOException {
    return OBJECT_MAPPER.readValue(fixturePath(fixture).toFile(), modelType);
  }

  private static Path fixturePath(String fixture) {
    return Path.of("..", "spec", "fixtures", fixture);
  }

  private record TransitionFixture(@JsonProperty("cases") List<TransitionCase> cases) {}

  private record TransitionCase(
      @JsonProperty("name") String name,
      @JsonProperty("from") String from,
      @JsonProperty("to") String to,
      @JsonProperty("pending_human_approval") boolean pendingHumanApproval,
      @JsonProperty("actor_is_participant") boolean actorIsParticipant,
      @JsonProperty("valid") boolean valid) {}
}
